File: src/main/java/komu/blunt/parser/OperatorSet.java
package komu.blunt.parser;

import java.util.HashMap;
import java.util.Map;

import static com.google.common.base.Preconditions.checkArgument;
import static java.lang.Math.max;

public final class OperatorSet {
    
    private final Map<String,Operator> ops = new HashMap<>();
    private int maxPrecedence = 0;
    private static final int DEFAULT_PRECEDENCE = 0;
    
    public OperatorSet() {
        add(1, Associativity.RIGHT, ":");
        add(2, "==", "<", "<=", ">", ">=");
        add(3, "+", "-");
        add(4, "*", "/", "%");
    }

    public void add(int precedence, String... names) {
        add(precedence, Associativity.LEFT, names);
    }
    
    public void add(int precedence, Associativity associativity, String... names) {
        checkArgument(precedence >= 0);

        maxPrecedence = max(maxPrecedence, precedence);
        for (String name : names)
            ops.put(name, new Operator(name, associativity, precedence));
    }

    public Operator operator(String name) {
        Operator op = ops.get(name);
        if (op != null)
            return op;
        else
            return new Operator(name, Associativity.LEFT, DEFAULT_PRECEDENCE);
    }

    public int getMaxLevel() {
        return maxPrecedence;
    }
}

File: src/main/kotlin/komu/blunt/parser/OperatorSet.kt
package komu.blunt.parser

import java.lang.Math.max
import java.util.*

class OperatorSet() {

    private val ops = HashMap<String,Operator>()
    private var maxPrecedence = 0
    private val DEFAULT_PRECEDENCE = 0

    init {
        add(1, Associativity.RIGHT, "$")
        add(2, Associativity.RIGHT, ":")
        add(3, "==", "<", "<=", ">", ">=")
        add(4, "+", "-")
        add(5, "*", "/", "%")
    }

    fun add(precedence: Int, vararg names: String) {
        addInternal(precedence, Associativity.LEFT, names)
    }

    fun add(precedence: Int, associativity: Associativity, vararg names: String) {
        addInternal(precedence, associativity, names)
    }

    fun addInternal(precedence: Int, associativity: Associativity, names: Array<out String>) {
        require(precedence >= 0)

        maxPrecedence = max(maxPrecedence, precedence)
        for (name in names)
            ops[name] = Operator(name, associativity, precedence)
    }

    operator fun get(name: String): Operator =
        ops[name] ?: Operator(name, Associativity.LEFT, DEFAULT_PRECEDENCE)

    val maxLevel: Int
        get() = maxPrecedence
}
